sql: mismatched type panic during apply join

Running the following produces a panic:

```
SELECT (SELECT ARRAY (SELECT c.relname FROM pg_inherits AS i JOIN pg_class AS c ON c.oid = i.inhparent WHERE i.inhrelid = rel.oid ORDER BY inhseqno)) FROM pg_class AS rel;

panic: invalid datum type given: int[], expected name[] [recovered]
        panic: panic while executing 1 statements: SELECT (SELECT ARRAY (SELECT _._ FROM _ AS _ JOIN _ AS _ ON _._ = _._ WHERE _._ = _._ ORDER BY _)) FROM _ AS _; caused by invalid datum type given: int[], expected name[]

goroutine 503 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00365ed80, 0x7a25e00, 0xc0036d4d00, 0x6c8d000, 0xc006f48cb0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00365ed80, 0x7a25e00, 0xc0036d4d00)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x6c8d000, 0xc006f48cb0)
        /usr/local/Cellar/go/1.12.4/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc0067fd180, 0x7a7cec0, 0xc006bcbe60, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:167 +0x1e2
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc006b78f00, 0xc000039b00, 0x7a28280, 0xc003fdaeb0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:179 +0x517
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x7a25ec0, 0xc006fe1080, 0x7a31600, 0xc006b78f00, 0x7a04380, 0xc00098a700)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc006b78f00, 0x7a25ec0, 0xc006fe1080)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00396fa40, 0x7a25ec0, 0xc006fe1080, 0x73320a8, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000a62a20, 0xc0082e5aa0, 0xc003a3e6c0, 0xc003f49330, 0xc003808000, 0xc003991800, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000a62a20, 0x7a25ec0, 0xc006fe0930, 0xc003991800, 0xc0082e5aa0, 0xc003a3e6c0, 0x7a28280, 0xc003fdaeb0, 0xc003808000)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).runRightSidePlan(0xc0037f1e00, 0x7a25ec0, 0xc006fe0930, 0xc0032a5400, 0xc000143600, 0xc003791a40, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:328 +0x5de
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).Next(0xc0037f1e00, 0x7a25ec0, 0xc006fe0930, 0xc0032a5400, 0xc000143600, 0x0, 0x7a25ec0, 0xc006fe0930)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:273 +0x6bf
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc006b78500, 0xc000143600, 0x7a26f40, 0xc0037f1e00, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:171 +0x43c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x7a25ec0, 0xc006fe0930, 0x7a31600, 0xc006b78500, 0x7a04380, 0xc0035d3c00)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc006b78500, 0x7a25ec0, 0xc006fe0930)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00396f4a0, 0x7a25ec0, 0xc006fe0930, 0x73320a8, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000a62a20, 0xc0082e51a0, 0xc003a3e6c0, 0xc003f49dd0, 0xc00038b600, 0xc0032a4e00, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000a62a20, 0x7a25ec0, 0xc006fe3020, 0xc0032a4e00, 0xc0082e51a0, 0xc003a3e6c0, 0x7a26f40, 0xc0037f1e00, 0xc00038b600)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).runRightSidePlan(0xc0037f1200, 0x7a25ec0, 0xc006fe3020, 0xc0037f1400, 0xc00365f0d0, 0xc003791800, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:328 +0x5de
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).Next(0xc0037f1200, 0x7a25ec0, 0xc006fe3020, 0xc0037f1400, 0xc00365f0d0, 0x0, 0x7a25ec0, 0xc006fe3020)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:273 +0x6bf
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc000657900, 0xc00365f0d0, 0x7a26f40, 0xc0037f1200, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:171 +0x43c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x7a25ec0, 0xc006fe3020, 0x7a31600, 0xc000657900, 0x7a04380, 0xc000b7ce00)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000657900, 0x7a25ec0, 0xc006fe3020)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00396ed20, 0x7a25ec0, 0xc006fe3020, 0x73320a8, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000a62a20, 0xc0082e4ae0, 0xc003a3e6c0, 0xc003f4a870, 0xc00038b340, 0xc00365f1a8, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000a62a20, 0x7a25ec0, 0xc006bcbb00, 0xc00365f1a8, 0xc0082e4ae0, 0xc003a3e6c0, 0x7a27bc0, 0xc0035c07e0, 0xc00038b340)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
```

The `EXPLAIN(opt,types)` output is:

```
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
  project
   ├── columns: array:61(name[])
   ├── stats: [rows=1000]
   ├── cost: 346.042712
   ├── prune: (61)
   ├── inner-join-apply
   │    ├── columns: oid:1(oid) relname:2(name) relnamespace:3(oid) reltype:4(oid) relowner:5(oid) relam:6(oid) relfilenode:7(oid) reltablespace:8(oid) relpages:9(int) reltuples:10(float) relallvisible:11(int) reltoastrelid:12(oid) relhasindex:13(bool) relisshared:14(bool) relpersistence:15(char) relistemp:16(bool) relkind:17(char) relnatts:18(int) relchecks:19(int) relhasoids:20(bool) relhaspkey:21(bool) relhasrules:22(bool) relhastriggers:23(bool) relhassubclass:24(bool) relfrozenxid:25(int) relacl:26(string[]) reloptions:27(string[]) array_agg:58(int[]) array:59(name[])
   │    ├── stats: [rows=1000]
   │    ├── cost: 326.032712
   │    ├── virtual-scan defaultdb.pg_catalog.pg_class
   │    │    ├── columns: oid:1(oid) relname:2(name) relnamespace:3(oid) reltype:4(oid) relowner:5(oid) relam:6(oid) relfilenode:7(oid) reltablespace:8(oid) relpages:9(int) reltuples:10(float) relallvisible:11(int) reltoastrelid:12(oid) relhasindex:13(bool) relisshared:14(bool) relpersistence:15(char) relistemp:16(bool) relkind:17(char) relnatts:18(int) relchecks:19(int) relhasoids:20(bool) relhaspkey:21(bool) relhasrules:22(bool) relhastriggers:23(bool) relhassubclass:24(bool) relfrozenxid:25(int) relacl:26(string[]) reloptions:27(string[])
   │    │    ├── stats: [rows=1000]
   │    │    └── cost: 10.01
   │    ├── inner-join-apply
   │    │    ├── columns: array_agg:58(int[]) array:59(name[])
   │    │    ├── outer: (1)
   │    │    ├── cardinality: [1 - 1]
   │    │    ├── stats: [rows=1]
   │    │    ├── cost: 293.495212
   │    │    ├── key: ()
   │    │    ├── fd: ()-->(58,59)
   │    │    ├── prune: (59)
   │    │    ├── project
   │    │    │    ├── columns: array_agg:58(int[])
   │    │    │    ├── outer: (1)
   │    │    │    ├── cardinality: [1 - 1]
   │    │    │    ├── stats: [rows=1]
   │    │    │    ├── cost: 293.425212
   │    │    │    ├── key: ()
   │    │    │    ├── fd: ()-->(58)
   │    │    │    ├── prune: (58)
   │    │    │    ├── group-by
   │    │    │    │    ├── columns: inhrelid:28(oid) array_agg:60(int[])
   │    │    │    │    ├── internal-ordering: +30 opt(28)
   │    │    │    │    ├── outer: (1)
   │    │    │    │    ├── cardinality: [1 - 1]
   │    │    │    │    ├── stats: [rows=1]
   │    │    │    │    ├── cost: 293.395212
   │    │    │    │    ├── key: ()
   │    │    │    │    ├── fd: ()-->(28,60)
   │    │    │    │    ├── sort
   │    │    │    │    │    ├── columns: inhrelid:28(oid) inhparent:29(oid) inhseqno:30(int) oid:31(oid) relname:32(name) relnamespace:33(oid) reltype:34(oid) relowner:35(oid) relam:36(oid) relfilenode:37(oid) reltablespace:38(oid) relpages:39(int) reltuples:40(float) relallvisible:41(int) reltoastrelid:42(oid) relhasindex:43(bool) relisshared:44(bool) relpersistence:45(char) relistemp:46(bool) relkind:47(char) relnatts:48(int) relchecks:49(int) relhasoids:50(bool) relhaspkey:51(bool) relhasrules:52(bool) relhastriggers:53(bool) relhassubclass:54(bool) relfrozenxid:55(int) relacl:56(string[]) reloptions:57(string[])
   │    │    │    │    │    ├── outer: (1)
   │    │    │    │    │    ├── cardinality: [1 - ]
   │    │    │    │    │    ├── stats: [rows=100, distinct(1)=1, null(1)=0, distinct(28)=1, null(28)=0]
   │    │    │    │    │    ├── cost: 291.375212
   │    │    │    │    │    ├── fd: (29)==(31), (31)==(29)
   │    │    │    │    │    ├── ordering: +30 opt(28) [actual: +30]
   │    │    │    │    │    └── right-join (hash)
   │    │    │    │    │         ├── columns: inhrelid:28(oid) inhparent:29(oid) inhseqno:30(int) oid:31(oid) relname:32(name) relnamespace:33(oid) reltype:34(oid) relowner:35(oid) relam:36(oid) relfilenode:37(oid) reltablespace:38(oid) relpages:39(int) reltuples:40(float) relallvisible:41(int) reltoastrelid:42(oid) relhasindex:43(bool) relisshared:44(bool) relpersistence:45(char) relistemp:46(bool) relkind:47(char) relnatts:48(int) relchecks:49(int) relhasoids:50(bool) relhaspkey:51(bool) relhasrules:52(bool) relhastriggers:53(bool) relhassubclass:54(bool) relfrozenxid:55(int) relacl:56(string[]) reloptions:57(string[])
   │    │    │    │    │         ├── outer: (1)
   │    │    │    │    │         ├── cardinality: [1 - ]
   │    │    │    │    │         ├── stats: [rows=100, distinct(1)=1, null(1)=0, distinct(28)=1, null(28)=0]
   │    │    │    │    │         ├── cost: 276.0775
   │    │    │    │    │         ├── fd: (29)==(31), (31)==(29)
   │    │    │    │    │         ├── inner-join (hash)
   │    │    │    │    │         │    ├── columns: inhrelid:28(oid) inhparent:29(oid!null) inhseqno:30(int) oid:31(oid!null) relname:32(name) relnamespace:33(oid) reltype:34(oid) relowner:35(oid) relam:36(oid) relfilenode:37(oid) reltablespace:38(oid) relpages:39(int) reltuples:40(float) relallvisible:41(int) reltoastrelid:42(oid) relhasindex:43(bool) relisshared:44(bool) relpersistence:45(char) relistemp:46(bool) relkind:47(char) relnatts:48(int) relchecks:49(int) relhasoids:50(bool) relhaspkey:51(bool) relhasrules:52(bool) relhastriggers:53(bool) relhassubclass:54(bool) relfrozenxid:55(int) relacl:56(string[]) reloptions:57(string[])
   │    │    │    │    │         │    ├── stats: [rows=10000, distinct(28)=100, null(28)=100, distinct(29)=100, null(29)=0, distinct(31)=100, null(31)=0]
   │    │    │    │    │         │    ├── cost: 150.03
   │    │    │    │    │         │    ├── fd: (29)==(31), (31)==(29)
   │    │    │    │    │         │    ├── virtual-scan defaultdb.pg_catalog.pg_inherits
   │    │    │    │    │         │    │    ├── columns: inhrelid:28(oid) inhparent:29(oid) inhseqno:30(int)
   │    │    │    │    │         │    │    ├── stats: [rows=1000, distinct(28)=100, null(28)=10, distinct(29)=100, null(29)=10]
   │    │    │    │    │         │    │    └── cost: 10.01
   │    │    │    │    │         │    ├── virtual-scan defaultdb.pg_catalog.pg_class
   │    │    │    │    │         │    │    ├── columns: oid:31(oid) relname:32(name) relnamespace:33(oid) reltype:34(oid) relowner:35(oid) relam:36(oid) relfilenode:37(oid) reltablespace:38(oid) relpages:39(int) reltuples:40(float) relallvisible:41(int) reltoastrelid:42(oid) relhasindex:43(bool) relisshared:44(bool) relpersistence:45(char) relistemp:46(bool) relkind:47(char) relnatts:48(int) relchecks:49(int) relhasoids:50(bool) relhaspkey:51(bool) relhasrules:52(bool) relhastriggers:53(bool) relhassubclass:54(bool) relfrozenxid:55(int) relacl:56(string[]) reloptions:57(string[])
   │    │    │    │    │         │    │    ├── stats: [rows=1000, distinct(31)=100, null(31)=10]
   │    │    │    │    │         │    │    └── cost: 10.01
   │    │    │    │    │         │    └── filters
   │    │    │    │    │         │         └── eq [type=bool, outer=(29,31), constraints=(/29: (/NULL - ]; /31: (/NULL - ]), fd=(29)==(31), (31)==(29)]
   │    │    │    │    │         │              ├── variable: oid [type=oid]
   │    │    │    │    │         │              └── variable: inhparent [type=oid]
   │    │    │    │    │         ├── values
   │    │    │    │    │         │    ├── cardinality: [1 - 1]
   │    │    │    │    │         │    ├── stats: [rows=1]
   │    │    │    │    │         │    ├── cost: 0.02
   │    │    │    │    │         │    ├── key: ()
   │    │    │    │    │         │    └── tuple [type=tuple]
   │    │    │    │    │         └── filters
   │    │    │    │    │              └── eq [type=bool, outer=(1,28), constraints=(/1: (/NULL - ]; /28: (/NULL - ]), fd=(1)==(28), (28)==(1)]
   │    │    │    │    │                   ├── variable: inhrelid [type=oid]
   │    │    │    │    │                   └── variable: oid [type=oid]
   │    │    │    │    └── aggregations
   │    │    │    │         ├── array-agg [type=int[], outer=(30)]
   │    │    │    │         │    └── variable: inhseqno [type=int]
   │    │    │    │         └── any-not-null-agg [type=oid, outer=(28)]
   │    │    │    │              └── variable: inhrelid [type=oid]
   │    │    │    └── projections
   │    │    │         └── case [type=int[], outer=(28,60)]
   │    │    │              ├── true [type=bool]
   │    │    │              ├── when [type=int[]]
   │    │    │              │    ├── is-not [type=bool]
   │    │    │              │    │    ├── variable: inhrelid [type=oid]
   │    │    │              │    │    └── null [type=unknown]
   │    │    │              │    └── variable: array_agg [type=int[]]
   │    │    │              └── null [type=unknown]
   │    │    ├── values
   │    │    │    ├── columns: array:59(name[])
   │    │    │    ├── outer: (58)
   │    │    │    ├── cardinality: [1 - 1]
   │    │    │    ├── stats: [rows=1]
   │    │    │    ├── cost: 0.02
   │    │    │    ├── key: ()
   │    │    │    ├── fd: ()-->(59)
   │    │    │    ├── prune: (59)
   │    │    │    └── tuple [type=tuple{int[]}]
   │    │    │         └── coalesce [type=int[]]
   │    │    │              ├── variable: array_agg [type=int[]]
   │    │    │              └── const: ARRAY[] [type=int[]]
   │    │    └── filters (true)
   │    └── filters (true)
   └── projections
        └── variable: array [type=name[], outer=(59)]
(113 rows)

Time: 3.032ms
```

It looks like the outer apply join is maybe returning the columns in the wrong order - I'm not really too sure yet, but just filing this for now.

--- a/pkg/sql/logictest/testdata/logic_test/subquery_correlated
+++ b/pkg/sql/logictest/testdata/logic_test/subquery_correlated
@@ -1150,3 +1150,23 @@ ORDER BY
 statement ok
 DROP TABLE stuff;
 DROP TABLE users;
+
+# Regression test for #38867.
+query T
+SELECT (
+		SELECT
+			ARRAY (
+			  SELECT c.relname
+			  FROM pg_inherits AS i JOIN pg_class AS c ON c.oid = i.inhparent
+			  WHERE i.inhrelid = rel.oid
+			  ORDER BY inhseqno
+			)
+)
+FROM pg_class AS rel
+LIMIT 5;
+----
+{}
+{}
+{}
+{}
+{}

--- a/pkg/sql/opt/colset.go
+++ b/pkg/sql/opt/colset.go
@@ -10,7 +10,10 @@
 
 package opt
 
-import "github.com/cockroachdb/cockroach/pkg/util"
+import (
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/errors"
+)
 
 // ColSet efficiently stores an unordered set of column ids.
 type ColSet struct {
@@ -85,3 +88,13 @@ func (s ColSet) SubsetOf(rhs ColSet) bool { return s.set.SubsetOf(rhs.set) }
 // numbers are shown as ranges. For example, for the set {1, 2, 3  5, 6, 10},
 // the output is "(1-3,5,6,10)".
 func (s ColSet) String() string { return s.set.String() }
+
+// SingleColumn returns the single column in s. Panics if s does not contain
+// exactly one column.
+func (s ColSet) SingleColumn() ColumnID {
+	if s.Len() != 1 {
+		panic(errors.AssertionFailedf("expected a single column but found %d columns", s.Len()))
+	}
+	col, _ := s.Next(0)
+	return col
+}

--- a/pkg/sql/opt/exec/factory.go
+++ b/pkg/sql/opt/exec/factory.go
@@ -104,7 +104,8 @@ type Factory interface {
 	//
 	// memo, rightProps, and right are the memo, required physical properties, and
 	// RelExpr of the right side of the join that will be repeatedly modified,
-	// re-planned and executed for every row from the left side.
+	// re-planned and executed for every row from the left side. The rightProps
+	// always includes a presentation.
 	//
 	// fakeRight is a pre-planned node that is the right side of the join with
 	// all outer columns replaced by NULL. The physical properties of this node

--- a/pkg/sql/opt/memo/typing.go
+++ b/pkg/sql/opt/memo/typing.go
@@ -249,7 +249,7 @@ func typeCollate(e opt.ScalarExpr) *types.T {
 // typeArrayFlatten returns the type of the subquery as an array.
 func typeArrayFlatten(e opt.ScalarExpr) *types.T {
 	input := e.Child(0).(RelExpr)
-	colID, _ := input.Relational().OutputCols.Next(0)
+	colID := e.(*ArrayFlattenExpr).RequestedCol
 	return types.MakeArray(input.Memo().Metadata().ColumnMeta(colID).Type)
 }
 
@@ -354,7 +354,7 @@ func typeCast(e opt.ScalarExpr) *types.T {
 // its first (and only) column.
 func typeSubquery(e opt.ScalarExpr) *types.T {
 	input := e.Child(0).(RelExpr)
-	colID, _ := input.Relational().OutputCols.Next(0)
+	colID := input.Relational().OutputCols.SingleColumn()
 	return input.Memo().Metadata().ColumnMeta(colID).Type
 }
 

--- a/pkg/sql/opt/norm/custom_funcs.go
+++ b/pkg/sql/opt/norm/custom_funcs.go
@@ -156,10 +156,9 @@ func (c *CustomFuncs) CanConstructBinary(op opt.Operator, left, right opt.Scalar
 	return memo.BinaryOverloadExists(op, left.DataType(), right.DataType())
 }
 
-// ArrayType returns the type of the first output column wrapped
+// ArrayType returns the type of the given column wrapped
 // in an array.
-func (c *CustomFuncs) ArrayType(in memo.RelExpr) *types.T {
-	inCol, _ := c.OutputCols(in).Next(0)
+func (c *CustomFuncs) ArrayType(inCol opt.ColumnID) *types.T {
 	inTyp := c.mem.Metadata().ColumnMeta(inCol).Type
 	return types.MakeArray(inTyp)
 }
@@ -1626,10 +1625,10 @@ func (c *CustomFuncs) SubqueryOrdering(sub *memo.SubqueryPrivate) physical.Order
 	return oc
 }
 
-// FirstCol returns the first column in the input expression.
-func (c *CustomFuncs) FirstCol(in memo.RelExpr) opt.ColumnID {
-	inCol, _ := c.OutputCols(in).Next(0)
-	return inCol
+// SubqueryRequestedCol returns the requested column from a SubqueryPrivate.
+// This function should only be used with ArrayFlatten expressions.
+func (c *CustomFuncs) SubqueryRequestedCol(sub *memo.SubqueryPrivate) opt.ColumnID {
+	return sub.RequestedCol
 }
 
 // MakeArrayAggCol returns a ColPrivate with the given type and an "array_agg" label.

--- a/pkg/sql/opt/norm/decorrelate.go
+++ b/pkg/sql/opt/norm/decorrelate.go
@@ -697,11 +697,7 @@ func (c *CustomFuncs) ConstructNoColsRow() memo.RelExpr {
 // referenceSingleColumn returns a Variable operator that refers to the one and
 // only column that is projected by the input expression.
 func (c *CustomFuncs) referenceSingleColumn(in memo.RelExpr) opt.ScalarExpr {
-	cols := in.Relational().OutputCols
-	if cols.Len() != 1 {
-		panic(errors.AssertionFailedf("expression does not have exactly one column"))
-	}
-	colID, _ := cols.Next(0)
+	colID := in.Relational().OutputCols.SingleColumn()
 	return c.f.ConstructVariable(colID)
 }
 
@@ -801,8 +797,14 @@ func (r *subqueryHoister) hoistAll(scalar opt.ScalarExpr) opt.ScalarExpr {
 		}
 
 		// Replace the Subquery operator with a Variable operator referring to
-		// the first (and only) column in the hoisted query.
-		colID, _ := subqueryProps.OutputCols.Next(0)
+		// the output column of the hoisted query.
+		var colID opt.ColumnID
+		switch t := scalar.(type) {
+		case *memo.ArrayFlattenExpr:
+			colID = t.RequestedCol
+		default:
+			colID = subqueryProps.OutputCols.SingleColumn()
+		}
 		return r.f.ConstructVariable(colID)
 	}
 

--- a/pkg/sql/opt/norm/rules/scalar.opt
+++ b/pkg/sql/opt/norm/rules/scalar.opt
@@ -288,8 +288,8 @@
                 $input
                 [
                     (AggregationsItem
-                        (ArrayAgg (Variable (FirstCol $input)))
-                        (MakeArrayAggCol (ArrayType $input))
+                        (ArrayAgg (Variable $requestedCol:(SubqueryRequestedCol $subquery)))
+                        (MakeArrayAggCol (ArrayType $requestedCol))
                     )
                 ]
                 (MakeOrderedGrouping
@@ -299,6 +299,6 @@
             )
             (MakeUnorderedSubquery)
         )
-        (Array [] (ArrayType $input))
+        (Array [] (ArrayType $requestedCol))
     ]
 )

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -1287,3 +1287,127 @@ project
            └── scan a
                 ├── columns: k:7(int!null)
                 └── key: (7)
+
+exec-ddl
+CREATE TABLE pg_class (
+     oid OID NULL,
+     relname NAME NOT NULL,
+     relnamespace OID NULL,
+     reltype OID NULL,
+     reloftype OID NULL,
+     relowner OID NULL,
+     relam OID NULL,
+     relfilenode OID NULL,
+     reltablespace OID NULL,
+     relpages INT4 NULL,
+     reltuples FLOAT4 NULL,
+     relallvisible INT4 NULL,
+     reltoastrelid OID NULL,
+     relhasindex BOOL NULL,
+     relisshared BOOL NULL,
+     relpersistence CHAR NULL,
+     relistemp BOOL NULL,
+     relkind CHAR NULL,
+     relnatts INT2 NULL,
+     relchecks INT2 NULL,
+     relhasoids BOOL NULL,
+     relhaspkey BOOL NULL,
+     relhasrules BOOL NULL,
+     relhastriggers BOOL NULL,
+     relhassubclass BOOL NULL,
+     relfrozenxid INT8 NULL,
+     relacl STRING[] NULL,
+     reloptions STRING[] NULL
+)
+----
+
+exec-ddl
+CREATE TABLE pg_inherits (
+      inhrelid OID NULL,
+      inhparent OID NULL,
+      inhseqno INT4 NULL
+)
+----
+
+# Regression test for #38867.
+norm expect=NormalizeArrayFlattenToAgg
+SELECT (
+		SELECT
+			ARRAY (
+			  SELECT c.relname
+			  FROM pg_inherits AS i JOIN pg_class AS c ON c.oid = i.inhparent
+			  WHERE i.inhrelid = rel.oid
+			  ORDER BY inhseqno
+			)
+)
+FROM pg_class AS rel
+----
+project
+ ├── columns: array:66(name[])
+ ├── inner-join-apply
+ │    ├── columns: rel.oid:1(oid) array_agg:63(name[]) array:64(name[])
+ │    ├── scan rel
+ │    │    └── columns: rel.oid:1(oid)
+ │    ├── inner-join-apply
+ │    │    ├── columns: array_agg:63(name[]) array:64(name[])
+ │    │    ├── outer: (1)
+ │    │    ├── cardinality: [1 - 1]
+ │    │    ├── key: ()
+ │    │    ├── fd: ()-->(63,64)
+ │    │    ├── project
+ │    │    │    ├── columns: array_agg:63(name[])
+ │    │    │    ├── outer: (1)
+ │    │    │    ├── cardinality: [1 - 1]
+ │    │    │    ├── key: ()
+ │    │    │    ├── fd: ()-->(63)
+ │    │    │    ├── group-by
+ │    │    │    │    ├── columns: inhrelid:30(oid) array_agg:65(name[])
+ │    │    │    │    ├── internal-ordering: +32 opt(30)
+ │    │    │    │    ├── outer: (1)
+ │    │    │    │    ├── cardinality: [1 - 1]
+ │    │    │    │    ├── key: ()
+ │    │    │    │    ├── fd: ()-->(30,65)
+ │    │    │    │    ├── sort
+ │    │    │    │    │    ├── columns: inhrelid:30(oid) inhparent:31(oid) inhseqno:32(int4) c.oid:34(oid) c.relname:35(name)
+ │    │    │    │    │    ├── outer: (1)
+ │    │    │    │    │    ├── cardinality: [1 - ]
+ │    │    │    │    │    ├── fd: (31)==(34), (34)==(31)
+ │    │    │    │    │    ├── ordering: +32 opt(30) [actual: +32]
+ │    │    │    │    │    └── left-join (hash)
+ │    │    │    │    │         ├── columns: inhrelid:30(oid) inhparent:31(oid) inhseqno:32(int4) c.oid:34(oid) c.relname:35(name)
+ │    │    │    │    │         ├── outer: (1)
+ │    │    │    │    │         ├── cardinality: [1 - ]
+ │    │    │    │    │         ├── fd: (31)==(34), (34)==(31)
+ │    │    │    │    │         ├── values
+ │    │    │    │    │         │    ├── cardinality: [1 - 1]
+ │    │    │    │    │         │    ├── key: ()
+ │    │    │    │    │         │    └── tuple [type=tuple]
+ │    │    │    │    │         ├── inner-join (hash)
+ │    │    │    │    │         │    ├── columns: inhrelid:30(oid) inhparent:31(oid!null) inhseqno:32(int4) c.oid:34(oid!null) c.relname:35(name!null)
+ │    │    │    │    │         │    ├── fd: (31)==(34), (34)==(31)
+ │    │    │    │    │         │    ├── scan i
+ │    │    │    │    │         │    │    └── columns: inhrelid:30(oid) inhparent:31(oid) inhseqno:32(int4)
+ │    │    │    │    │         │    ├── scan c
+ │    │    │    │    │         │    │    └── columns: c.oid:34(oid) c.relname:35(name!null)
+ │    │    │    │    │         │    └── filters
+ │    │    │    │    │         │         └── c.oid = inhparent [type=bool, outer=(31,34), constraints=(/31: (/NULL - ]; /34: (/NULL - ]), fd=(31)==(34), (34)==(31)]
+ │    │    │    │    │         └── filters
+ │    │    │    │    │              └── inhrelid = rel.oid [type=bool, outer=(1,30), constraints=(/1: (/NULL - ]; /30: (/NULL - ]), fd=(1)==(30), (30)==(1)]
+ │    │    │    │    └── aggregations
+ │    │    │    │         ├── array-agg [type=name[], outer=(35)]
+ │    │    │    │         │    └── variable: c.relname [type=name]
+ │    │    │    │         └── any-not-null-agg [type=oid, outer=(30)]
+ │    │    │    │              └── variable: inhrelid [type=oid]
+ │    │    │    └── projections
+ │    │    │         └── CASE WHEN inhrelid IS NOT NULL THEN array_agg END [type=name[], outer=(30,65)]
+ │    │    ├── values
+ │    │    │    ├── columns: array:64(name[])
+ │    │    │    ├── outer: (63)
+ │    │    │    ├── cardinality: [1 - 1]
+ │    │    │    ├── key: ()
+ │    │    │    ├── fd: ()-->(64)
+ │    │    │    └── (COALESCE(array_agg, ARRAY[]),) [type=tuple{name[]}]
+ │    │    └── filters (true)
+ │    └── filters (true)
+ └── projections
+      └── variable: array [type=name[], outer=(64)]

--- a/pkg/sql/opt/ops/scalar.opt
+++ b/pkg/sql/opt/ops/scalar.opt
@@ -42,7 +42,7 @@ define SubqueryPrivate {
 
     # RequestedCol is set if there could possibly be other columns in the input
     # (say, if there was an ordering that must be respected) besides the one that
-    # will eventually be output.
+    # will eventually be output. It is only used for ArrayFlatten expressions.
     RequestedCol ColumnID
 
     # Cmp is only used for AnyOp.

